crates/primitives/src/chain_spec.rs

```diff
@@ -0,0 +1,236 @@
+    chain_id: ChainId,
+    genesis: Genesis,
+    genesis_hash: H256,
+    hardforks: BTreeMap<Hardfork, BlockNumber>,
+    // This is an edge case, maybe we move it into a `Hardforks` struct
+    paris_block: Option<u64>,
+    paris_ttd: Option<U256>,
+    shanghai_block: Option<u64>,
+}
+
+impl ChainSpec {
+    /// Returns the Etereum mainnet spec
+    pub fn mainnet() -> Self {
+        Self {
+            chain_id: 1,
+            genesis: Genesis::default(),
+            genesis_hash: H256(hex!(
+                "d4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3"
+            )),
+            hardforks: BTreeMap::from([
+                (Hardfork::Frontier, 0),
+                (Hardfork::Homestead, 1150000),
+                (Hardfork::Dao, 1920000),
+                (Hardfork::Tangerine, 2463000),
+                (Hardfork::SpuriousDragon, 2675000),
+                (Hardfork::Byzantium, 4370000),
+                (Hardfork::Constantinople, 7280000),
+                (Hardfork::Petersburg, 7280000),
+                (Hardfork::Istanbul, 9069000),
+                (Hardfork::Muirglacier, 9200000),
+                (Hardfork::Berlin, 12244000),
+                (Hardfork::London, 12965000),
+                (Hardfork::ArrowGlacier, 13773000),
+                (Hardfork::GrayGlacier, 15050000),
+                (Hardfork::Latest, 15050000),
+            ]),
+            paris_block: Some(15537394),
+            paris_ttd: Some(U256::from(58750000000000000000000_u128)),
+            shanghai_block: None,
+        }
+    }
+
+    /// Returns the chain id
+    pub fn chain_id(&self) -> ChainId {
+        self.chain_id
+    }
+
+    /// Returns the chain genesis hash
+    pub fn genesis_hash(&self) -> H256 {
+        self.genesis_hash
+    }
+
+    /// Get the first block number of the hardfork.
+    pub fn fork_block(&self, fork: Hardfork) -> Option<BlockNumber> {
+        self.hardforks.get(&fork).copied()
+    }
+
+    /// Get the Paris status
+    pub fn paris_status(&self) -> ParisStatus {
+        match self.paris_ttd {
+            Some(terminal_total_difficulty) => {
+                ParisStatus::Supported { terminal_total_difficulty, block: self.paris_block }
+            }
+            None => ParisStatus::NotSupported,
+        }
+    }
+
+    /// Get the Shanghai block number
+    pub fn shanghai_block(&self) -> Option<BlockNumber> {
+        self.shanghai_block
+    }
+
+    /// Get an iterator of all harforks with theirs respectives block number
+    pub fn all_forks_blocks(&self) -> impl Iterator<Item = (Hardfork, BlockNumber)> + '_ {
+        self.hardforks.iter().map(|(f, b)| (*f, *b))
+    }
+
+    /// Creates a [`ForkFilter`](crate::ForkFilter) for the given hardfork.
+    ///
+    /// **CAUTION**: This assumes the current hardfork's block number is the current head and uses
+    /// all known future hardforks to initialize the filter.
+    pub fn fork_filter(&self, fork: Hardfork) -> Option<ForkFilter> {
```

Comment:
I think the networking component has access to the current head we've synced to because we emit status updates to it, so maybe a better API here would be to take a block number and resolve the filter that way? cc @mattsse